tsne/bh_sne_src/tsne.cpp

```diff
@@ -54,10 +56,65 @@ using std::fprintf;
 using std::log;
 using std::move;
 using std::sqrt;
+using std::unique_ptr;
 using std::vector;
 
 namespace {
 
+struct TSNEState {
+ public:
+  TSNEState(int N, double* const Y, int no_dims, double theta, int max_iter,
+            int stop_lying_iter, int mom_switch_iter, int iter,
+            float total_time, float residual_time, vector<double> P,
+            vector<unsigned int> row_P, vector<unsigned int> col_P,
+            vector<double> val_P, vector<double> dY, vector<double> uY,
+            vector<double> gains)
```

Comment:
Unless you're very careful, this is going to end up copying these vectors twice, once when they're passed to the constructor and then again when they're passed to the constructors for the members.  I think you can solve that problem with liberal application of `std::move`.